### ISSUE TITLE
Add a text asynchronous reader

### ISSUE THREAD

**Issue by szymonwieloch** [tokio-rs/tokio#3640]
Tokio has the io::BufReader struct for reading buffered bytes and the io::Lines struct for reading text lines asynchronously. Quite surprisingly it does not have any asynchronous reader that would asynchronously return pieces of UTF8 text once they are available. 

My code at the moment is reading data from a byte stream. There is no guarantee that this string will always contain valid UTF8 characters and also - because some UTF8 character consist of several bytes - you can get only a partial character. My code cannot wait for the end of line. I would like to be able to obtain parts of text as soon as they are available. It seems that io::Lines does exactly that but also adds line splitting. It would be possibly best to split Lines into two layers - one just parsing text and the second splitting text into lines.

Please notice that exactly this approach is used by this create:

https://docs.rs/textstream/0.1.1/textstream/

Unfortunately this implementation is synchronous and will not work in my case.

**Comment by sb64** [comment]
I'm interested in working on this. I'll post a suggested solution soon. I will note, however, that neither [`encoding`](https://docs.rs/encoding/0.2.33/encoding/index.html) (which `textstream` uses) nor [`encoding-rs`](https://docs.rs/encoding_rs/0.8.28/encoding_rs/index.html) are at 1.* yet, which means we can't expose them in a public API, and although we could use one of them in the implementation, that would require adding an extra dependency to tokio, which may not be desired.

**Comment by Darksonn** [comment]
I don't think we want to add new dependencies for this feature. It's relatively simple to implement. I would probably put it in `tokio_util::io`.

**Comment by sb64** [comment]
That was my thought, since a new dependency adds more overhead and UTF-8 is not too complex to parse. One thing I'm hung up on is whether such a solution should be conceptually (although probably not literally, since `Stream` is still unstable) a `Stream` of `char`s (or `Option<char>`s, for invalid UTF-8), or a `Stream` of `String`s, where each `String` is text that is yielded as soon as it is available. Although if this was in `tokio_util`, would it be fine to implement it as a `futures::Stream`? My other question is whether this should be implemented using `AsyncRead` or `AsyncBufRead`. I'm unsure of the exact properties, benefits, and drawbacks of both, though I do understand that buffered reading can be more efficient because it uses fewer syscalls. However, if the goal is to obtain bits of text as soon as possible, should it be implemented with `AsyncRead`, so it doesn't spend time collecting more data into a buffer?

**Comment by Darksonn** [comment]
I don't think a `Stream` is a good idea. I was imagining that you have an `AsyncRead` or `AsyncBufRead` that withholds a few bytes if it got half of an utf-8 character.

**Comment by sb64** [comment]
*Oh*, so we have some `struct TextReader<T: Async{Read, BufRead}>` that yields all bytes it receives from `T`, barring incomplete UTF-8 characters?

That seems better. Do you think it should be implemented as `AsyncRead` or `AsyncBufRead`? My gut reaction is `AsyncBufRead`, because with `AsyncRead`, I don't know how a type could inspect the bytes read to check for UTF-8 characters without having its own buffer, which would make it more of an `AsyncBufRead` anyways.

Another unresolved question is how to handle invalid UTF-8. I suppose if we go with a lower-level API like the one you suggested, it's up to the user whether to use `from_utf8` or `from_utf8_lossy`.

**Comment by luben** [comment]
> Another unresolved question is how to handle invalid UTF-8.

I think the right approach is to replace the invalid sequences with the unicode REPLACEMENT CHARACTER (for UTF-8 is 0xEF 0xBF 0xBD).

**Comment by sb64** [comment]
The issue with that is that I can't mutate the bytes obtained from the underlying reader, so I'd have to have my own buffer, which would be expensive. If I just dumbly forward invalid UTF-8 along, then the user can use `std::str::from_utf8_lossy` to do what you say, replace invalid bytes with the Unicode REPLACEMENT CHARACTER, use `std::str::from_utf8` to return an error or panic using `unwrap_*` or `expect`, or use `std::str::from_utf8_unchecked` if they're absolutely certain that it will be valid UTF-8.

Any thoughts Alice?
